core/android/common/src/test/kotlin/app/k9mail/core/android/common/contact/CachingContactRepositoryTest.kt

```diff
@@ -0,0 +1,113 @@
+package app.k9mail.core.android.common.contact
+
+import app.k9mail.core.common.cache.InMemoryCache
+import app.k9mail.core.common.mail.EmailAddress
+import assertk.assertThat
+import assertk.assertions.isEqualTo
+import assertk.assertions.isFalse
+import assertk.assertions.isNull
+import assertk.assertions.isTrue
+import kotlin.test.Test
+import org.junit.Before
+import org.junit.runner.RunWith
+import org.mockito.kotlin.doReturn
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.stub
+import org.robolectric.RobolectricTestRunner
+
+@RunWith(RobolectricTestRunner::class)
+internal class CachingContactRepositoryTest {
+
+    private val dataSource = mock<ContactDataSource>()
+    private val cache = InMemoryCache<EmailAddress, Contact>()
+
+    private val testSubject = CachingContactRepository(cache = cache, dataSource = dataSource)
+
+    @Before
+    fun setUp() {
+        cache.clear()
+    }
+
+    @Test
+    fun `getContactFor() returns null if no contact exists`() {
+        val result = testSubject.getContactFor(CONTACT_EMAIL_ADDRESS)
+
+        assertThat(result).isNull()
+    }
+
+    @Test
+    fun `getContactFor() returns contact if it exists`() {
+        dataSource.stub { on { getContactFor(CONTACT_EMAIL_ADDRESS) } doReturn CONTACT }
+
+        val result = testSubject.getContactFor(CONTACT_EMAIL_ADDRESS)
+
+        assertThat(result).isEqualTo(CONTACT)
+    }
+
+    @Test
+    fun `getContactFor() caches contact`() {
```

Comment:
We can test this by checking whether the returned references of successive `getContactFor()` calls are identical. Ideally tests only check behavior observable by callers, not internal state.